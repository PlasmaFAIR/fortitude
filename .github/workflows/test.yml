name: Tests

on:
  push:
    paths:
      - '**.rs'
      - '**/Cargo.toml'
      - '**/Cargo.lock'
      - '.github/workflows/test.yml'
  pull_request:
    paths:
      - '**.rs'
      - '**/Cargo.toml'
      - '**/Cargo.lock'
      - '.github/workflows/test.yml'
  workflow_dispatch:

env:
  CARGO_INCREMENTAL: 0
  CARGO_TERM_COLOR: always

jobs:

  build-linux:
    name: Build and Test
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - name: "Install cargo nextest"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-nextest
    - name: "Install cargo insta"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-insta
    - name: "Build"
      run: cargo build --verbose
    - name: "Run tests"
      shell: bash
      env:
        NEXTEST_PROFILE: "ci"
      run: cargo insta test --all-features --unreferenced reject --test-runner nextest
    - name: "Check docs"
      run: cargo dev generate-all --mode=check
    - name: "Check docs for broken links"
      run: cargo doc --all --no-deps
      env:
        # Setting RUSTDOCFLAGS because `cargo doc --check` isn't yet
        # implemented (https://github.com/rust-lang/cargo/issues/10025)
        RUSTDOCFLAGS: "-D warnings"
    - uses: actions/upload-artifact@v4
      with:
        name: fortitude
        path: target/debug/fortitude

  build-windows:
    name: Build and Test (windows)
    runs-on: windows-latest
    steps:
    - uses: actions/checkout@v4
    - name: "Install cargo nextest"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-nextest
    - name: "Install cargo insta"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-insta
    - name: "Build"
      run: cargo build --verbose
    - name: "Run tests"
      shell: bash
      env:
        NEXTEST_PROFILE: "ci"
      run: cargo insta test --all-features --unreferenced reject --test-runner nextest

  build-macos:
    name: Build and Test (macos)
    runs-on: macos-latest
    steps:
    - uses: actions/checkout@v4
    - name: "Install cargo nextest"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-nextest
    - name: "Install cargo insta"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-insta
    - name: "Build"
      run: cargo build --verbose
    - name: "Run tests"
      shell: bash
      env:
        NEXTEST_PROFILE: "ci"
      run: cargo insta test --all-features --unreferenced reject --test-runner nextest

  cargo-install:
    name: Test cargo installation
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - name: Check cargo install
      run: |
        cargo install --path fortitude
        fortitude --help

  pip-install:
    name: Test pip installation
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - name: Set up Python
      uses: actions/setup-python@v5
      with:
        python-version: '3.x'
    - name: Check pip install
      run: |
        pip install .
        fortitude --help
