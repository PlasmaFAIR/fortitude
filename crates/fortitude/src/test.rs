use std::path::Path;

use anyhow::Result;
use ruff_source_file::{SourceFile, SourceFileBuilder};

use crate::{
    check::{
        ast_entrypoint_map, check_file, read_to_string, rules_to_path_rules, rules_to_text_rules,
    },
    message::{Emitter, TextEmitter},
    rule_table::RuleTable,
    rules::Rule,
    settings::{self, FixMode, Settings},
};

#[macro_export]
macro_rules! apply_common_filters {
    {} => {
        let mut settings = insta::Settings::clone_current();
        // Convert windows paths to Unix Paths.
        settings.add_filter(r"\\\\?([\w\d.])", "/$1");
        let _bound = settings.bind_to_scope();
    }
}

pub(crate) fn test_resource_path(path: impl AsRef<Path>) -> std::path::PathBuf {
    Path::new("./resources/test/").join(path)
}

/// Run [`check_path`] on a Fortran file in the `resources/test/fixtures` directory.
pub(crate) fn test_path(
    path: impl AsRef<Path>,
    rules: &[Rule],
    settings: &Settings,
) -> Result<String> {
    let path = test_resource_path("fixtures").join(path);
    let source = read_to_string(&path)?;
    let filename = path.to_string_lossy();
    let file = SourceFileBuilder::new(filename.as_ref(), source.as_str()).finish();
    Ok(test_contents(&path, &file, rules, settings))
}

pub(crate) fn test_contents(
    path: &Path,
    file: &SourceFile,
    rules: &[Rule],
    settings: &Settings,
) -> String {
    let rule_table = RuleTable::from_iter(rules.iter().cloned());
    let path_rules = rules_to_path_rules(&rule_table);
    let text_rules = rules_to_text_rules(&rule_table);
    let ast_entrypoints = ast_entrypoint_map(&rule_table);

    match check_file(
        &rule_table,
        &path_rules,
        &text_rules,
        &ast_entrypoints,
        path,
        file,
        settings,
        FixMode::Generate,
        settings::IgnoreAllowComments::Disabled,
    ) {
        Ok(violations) => {
            if violations.messages.is_empty() {
                return String::new();
            }

            let mut output = Vec::new();

            TextEmitter::default()
                .with_show_fix_status(true)
                .with_show_fix_diff(true)
                .with_show_source(true)
                .with_unsafe_fixes(crate::settings::UnsafeFixes::Enabled)
                .emit(&mut output, &violations.messages)
                .unwrap();

            String::from_utf8(output).unwrap()
        }
        Err(msg) => {
            panic!("Failed to process: {msg}");
        }
    }
}
