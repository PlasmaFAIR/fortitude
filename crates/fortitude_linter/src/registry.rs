use fortitude_macros::RuleNamespace;
use rayon::str;
use std::str::FromStr; // Needed by strum_macros

pub use crate::rules::Rule;
pub use rule_set::{RuleSet, RuleSetIterator};

mod rule_set;

// Rule categories and identity codes
// ----------------------------------
// Helps to sort rules into logical categories, and acts as a unique tag with which
// users can switch rules on and off.

pub trait AsRule {
    fn rule(&self) -> Rule;
}

impl Rule {
    pub fn from_code(code: &str) -> Result<Self, FromCodeError> {
        let (category, code) = Category::parse_code(code).ok_or(FromCodeError::Unknown)?;
        category
            .all_rules()
            .find(|rule| rule.noqa_code().suffix() == code)
            .ok_or(FromCodeError::Unknown)
    }
}

#[derive(thiserror::Error, Debug)]
pub enum FromCodeError {
    #[error("unknown rule code")]
    Unknown,
}

/// The category of each rule defines the sort of problem it intends to solve.
#[derive(
    Debug,
    PartialEq,
    Eq,
    Clone,
    Hash,
    PartialOrd,
    Ord,
    strum_macros::AsRefStr,
    strum_macros::Display,
    strum_macros::EnumIter,
    strum_macros::EnumString,
    strum_macros::IntoStaticStr,
    RuleNamespace,
)]
#[repr(u16)]
#[strum(serialize_all = "kebab-case")]
pub enum Category {
    /// Failure to parse a file.
    #[prefix = "E"]
    Error,
    /// Detect code that is bug-prone or likely to be incorrect.
    #[prefix = "C"]
    Correctness,
    /// Obsolescent language features, as determined by the Fortran standard.
    #[prefix = "OB"]
    Obsolescent,
    /// Update to modern Fortran features. Used for less severe issues than `Obsolescent`,
    /// and goes beyond recommendations in the Fortran standard.
    #[prefix = "MOD"]
    Modernisation,
    /// Violation of style conventions.
    #[prefix = "S"]
    Style,
    /// Avoid platform/compiler-specific features.
    #[prefix = "PORT"]
    Portability,
    /// Fortitude specific rules.
    #[prefix = "FORT"]
    Fortitude,
}

pub trait RuleNamespace: Sized {
    /// Returns the prefix that every single code that fortitude uses to identify
    /// rules from this category starts with.
    fn common_prefix(&self) -> &'static str;

    /// Returns the variant corresponding to the given prefix.
    fn from_prefix(preifx: &str) -> Result<Self, &'static str>;

    /// Attempts to parse the given rule code. If the prefix is recognized
    /// returns the respective variant along with the code with the common
    /// prefix stripped.
    fn parse_code(code: &str) -> Option<(Self, &str)>;

    #[allow(dead_code)]
    fn name(&self) -> &'static str;

    #[allow(dead_code)]
    fn description(&self) -> &'static str;
}

pub mod clap_completion {
    use clap::builder::{PossibleValue, TypedValueParser, ValueParserFactory};
    use strum::IntoEnumIterator;

    use crate::registry::Rule;

    #[derive(Clone)]
    pub struct RuleParser;

    impl ValueParserFactory for Rule {
        type Parser = RuleParser;

        fn value_parser() -> Self::Parser {
            RuleParser
        }
    }

    impl TypedValueParser for RuleParser {
        type Value = Rule;

        fn parse_ref(
            &self,
            cmd: &clap::Command,
            arg: Option<&clap::Arg>,
            value: &std::ffi::OsStr,
        ) -> Result<Self::Value, clap::Error> {
            let value = value
                .to_str()
                .ok_or_else(|| clap::Error::new(clap::error::ErrorKind::InvalidUtf8))?;

            Rule::from_code(value).map_err(|_| {
                let mut error =
                    clap::Error::new(clap::error::ErrorKind::ValueValidation).with_cmd(cmd);
                if let Some(arg) = arg {
                    error.insert(
                        clap::error::ContextKind::InvalidArg,
                        clap::error::ContextValue::String(arg.to_string()),
                    );
                }
                error.insert(
                    clap::error::ContextKind::InvalidValue,
                    clap::error::ContextValue::String(value.to_string()),
                );
                error
            })
        }

        fn possible_values(&self) -> Option<Box<dyn Iterator<Item = PossibleValue> + '_>> {
            Some(Box::new(Rule::iter().map(|rule| {
                let name = rule.noqa_code().to_string();
                let help = rule.as_ref().to_string();
                PossibleValue::new(name).help(help)
            })))
        }
    }
}

#[cfg(test)]
mod tests {
    use std::mem::size_of;

    use strum::IntoEnumIterator;

    use super::{Category, Rule, RuleNamespace};

    #[test]
    fn documentation() {
        for rule in Rule::iter() {
            assert!(
                rule.explanation().is_some(),
                "Rule {} is missing documentation",
                rule.as_ref()
            );
        }
    }

    #[test]
    fn check_code_serialization() {
        for rule in Rule::iter() {
            assert!(
                Rule::from_code(&format!("{}", rule.noqa_code())).is_ok(),
                "{rule:?} could not be round-trip serialized."
            );
        }
    }

    #[test]
    fn category_parse_code() {
        for rule in Rule::iter() {
            let code = format!("{}", rule.noqa_code());
            let (category, rest) =
                Category::parse_code(&code).unwrap_or_else(|| panic!("couldn't parse {code:?}"));
            assert_eq!(code, format!("{}{rest}", category.common_prefix()));
        }
    }

    #[test]
    fn rule_size() {
        assert_eq!(2, size_of::<Rule>());
    }
}
