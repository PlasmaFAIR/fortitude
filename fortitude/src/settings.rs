// Some parts adapted from from ruff
// Copyright 2022 Charles Marsh
// SPDX-License-Identifier: MIT

/// A collection of user-modifiable settings. Should be expanded as new features are added.
use std::fmt;
use std::path::{Path, PathBuf};
use std::str::FromStr;

use lazy_static::lazy_static;
use path_absolutize::path_dedot;
use ruff_diagnostics::Applicability;
use ruff_macros::CacheKey;
use serde::{de, Deserialize, Deserializer, Serialize};
use strum::IntoEnumIterator;

use crate::display_settings;
use crate::fs::{FilePatternSet, EXCLUDE_BUILTINS, FORTRAN_EXTS};
use crate::registry::Rule;
use crate::rule_selector::{CompiledPerFileIgnoreList, PreviewOptions, RuleSelector};
use crate::rule_table::RuleTable;
use crate::rules::correctness::exit_labels;
use crate::rules::style::keywords;

#[derive(Debug)]
pub struct Settings {
    pub check: CheckSettings,
    pub file_resolver: FileResolverSettings,
}

impl Default for Settings {
    fn default() -> Self {
        let project_root = path_dedot::CWD.as_path();
        Self {
            check: CheckSettings::new(project_root),
            file_resolver: FileResolverSettings::new(project_root),
        }
    }
}

impl fmt::Display for Settings {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "\n# General Settings")?;
        display_settings! {
            formatter = f,
            fields = [
                self.check         | nested,
                self.file_resolver | nested,
            ]
        }
        Ok(())
    }
}

#[derive(Debug)]
pub struct CheckSettings {
    pub project_root: PathBuf,

    pub rules: RuleTable,
    pub per_file_ignores: CompiledPerFileIgnoreList,

    pub line_length: usize,

    pub fix: bool,
    pub fix_only: bool,
    pub show_fixes: bool,
    pub unsafe_fixes: UnsafeFixes,
    pub output_format: OutputFormat,
    pub progress_bar: ProgressBar,
    pub preview: PreviewMode,
    pub ignore_allow_comments: IgnoreAllowComments,

    // Individual rule settings
    pub exit_unlabelled_loops: exit_labels::settings::Settings,
    pub keyword_whitespace: keywords::settings::Settings,
}

impl CheckSettings {
    fn new(project_root: &Path) -> Self {
        Self {
            project_root: project_root.to_path_buf(),
            rules: DEFAULT_SELECTORS
                .iter()
                .flat_map(|selector| selector.rules(&PreviewOptions::default()))
                .collect(),
            per_file_ignores: CompiledPerFileIgnoreList::default(),
            line_length: 100,
            fix: false,
            fix_only: false,
            show_fixes: false,
            unsafe_fixes: UnsafeFixes::default(),
            output_format: OutputFormat::default(),
            progress_bar: ProgressBar::default(),
            preview: PreviewMode::default(),
            ignore_allow_comments: IgnoreAllowComments::default(),
            exit_unlabelled_loops: exit_labels::settings::Settings::default(),
            keyword_whitespace: keywords::settings::Settings::default(),
        }
    }
}

impl fmt::Display for CheckSettings {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "\n# Check Settings")?;
        display_settings! {
            formatter = f,
            namespace = "check",
            fields = [
                self.project_root | path,
                self.rules | nested,
                self.per_file_ignores,
                self.line_length,
                self.fix,
                self.fix_only,
                self.show_fixes,
                self.output_format,
                self.progress_bar,
                self.preview,
            ]
        }
        writeln!(f, "\n# Individual rules")?;
        display_settings! {
            formatter = f,
            namespace = "check",
            fields = [
                self.exit_unlabelled_loops | nested,
                self.keyword_whitespace | nested,
            ]
        }
        Ok(())
    }
}
#[derive(Debug, CacheKey)]
pub struct FileResolverSettings {
    pub excludes: FilePatternSet,
    pub force_exclude: bool,
    pub files: Vec<PathBuf>,
    pub file_extensions: Vec<String>,
    pub respect_gitignore: bool,
    pub project_root: PathBuf,
}

impl fmt::Display for FileResolverSettings {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "\n# File Resolver Settings")?;
        display_settings! {
            formatter = f,
            namespace = "file_resolver",
            fields = [
                self.excludes,
                self.force_exclude,
                self.files | paths,
                self.file_extensions | array,
                self.respect_gitignore,
                self.project_root | path,
            ]
        }
        Ok(())
    }
}

impl FileResolverSettings {
    fn new(project_root: &Path) -> Self {
        Self {
            project_root: project_root.to_path_buf(),
            excludes: FilePatternSet::try_from_iter(EXCLUDE_BUILTINS.iter().cloned()).unwrap(),
            force_exclude: false,
            respect_gitignore: true,
            files: Vec::default(),
            file_extensions: FORTRAN_EXTS.iter().map(|ext| ext.to_string()).collect(),
        }
    }
}

/// Toggle for rules still in preview
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, CacheKey, is_macro::Is)]
pub enum PreviewMode {
    #[default]
    Disabled,
    Enabled,
}

impl From<bool> for PreviewMode {
    fn from(version: bool) -> Self {
        if version {
            PreviewMode::Enabled
        } else {
            PreviewMode::Disabled
        }
    }
}

impl fmt::Display for PreviewMode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Disabled => write!(f, "disabled"),
            Self::Enabled => write!(f, "enabled"),
        }
    }
}

lazy_static! {
    /// Default rule selection
    pub static ref DEFAULT_SELECTORS: Vec<RuleSelector> = {
        Rule::iter()
            .filter(Rule::is_default)
            .map(|rule| {
                let code = rule.noqa_code();
                let code = format!("{}{}", code.prefix(), code.suffix());
                RuleSelector::from_str(code.as_str()).unwrap()
            })
            .collect()
    };
}

/// Toggle for unsafe fixes.
/// `Hint` will not apply unsafe fixes but a message will be shown when they are available.
/// `Disabled` will not apply unsafe fixes or show a message.
/// `Enabled` will apply unsafe fixes.
#[derive(Debug, Copy, Clone, CacheKey, Default, PartialEq, Eq, is_macro::Is)]
pub enum UnsafeFixes {
    #[default]
    Hint,
    Disabled,
    Enabled,
}

impl fmt::Display for UnsafeFixes {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::Hint => "hint",
                Self::Disabled => "disabled",
                Self::Enabled => "enabled",
            }
        )
    }
}

impl From<bool> for UnsafeFixes {
    fn from(value: bool) -> Self {
        if value {
            UnsafeFixes::Enabled
        } else {
            UnsafeFixes::Disabled
        }
    }
}

impl UnsafeFixes {
    pub fn required_applicability(&self) -> Applicability {
        match self {
            Self::Enabled => Applicability::Unsafe,
            Self::Disabled | Self::Hint => Applicability::Safe,
        }
    }
}

#[derive(Debug, Copy, Clone, Hash, Default, is_macro::Is)]
pub enum IgnoreAllowComments {
    Enabled,
    #[default]
    Disabled,
}

impl From<bool> for IgnoreAllowComments {
    fn from(value: bool) -> Self {
        if value {
            IgnoreAllowComments::Enabled
        } else {
            IgnoreAllowComments::Disabled
        }
    }
}

/// Command-line pattern for per-file rule ignores
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PatternPrefixPair {
    pub pattern: String,
    pub prefix: RuleSelector,
}

impl PatternPrefixPair {
    const EXPECTED_PATTERN: &'static str = "<FilePattern>:<RuleCode> pattern";
}

impl<'de> Deserialize<'de> for PatternPrefixPair {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let str_result = String::deserialize(deserializer)?;
        Self::from_str(str_result.as_str()).map_err(|_| {
            de::Error::invalid_value(
                de::Unexpected::Str(str_result.as_str()),
                &Self::EXPECTED_PATTERN,
            )
        })
    }
}

impl FromStr for PatternPrefixPair {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (pattern_str, code_string) = {
            let tokens = s.split(':').collect::<Vec<_>>();
            if tokens.len() != 2 {
                anyhow::bail!("Expected {}", Self::EXPECTED_PATTERN);
            }
            (tokens[0].trim(), tokens[1].trim())
        };
        let pattern = pattern_str.into();
        let prefix = RuleSelector::from_str(code_string)?;
        Ok(Self { pattern, prefix })
    }
}

/// Toggle for excluding files even when passed directly on the command line
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, CacheKey, is_macro::Is)]
pub enum ExcludeMode {
    #[default]
    NoForce,
    Force,
}

impl From<bool> for ExcludeMode {
    fn from(b: bool) -> Self {
        if b {
            Self::Force
        } else {
            Self::NoForce
        }
    }
}

/// Toggle for respecting .gitignore files
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, CacheKey, is_macro::Is)]
pub enum GitignoreMode {
    #[default]
    RespectGitignore,
    NoRespectGitignore,
}

impl From<bool> for GitignoreMode {
    fn from(value: bool) -> Self {
        if value {
            GitignoreMode::RespectGitignore
        } else {
            GitignoreMode::NoRespectGitignore
        }
    }
}

impl From<GitignoreMode> for bool {
    fn from(value: GitignoreMode) -> Self {
        match value {
            GitignoreMode::RespectGitignore => true,
            GitignoreMode::NoRespectGitignore => false,
        }
    }
}

/// Toggle for progress bar
#[derive(
    Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Debug, Hash, Default, clap::ValueEnum,
)]
#[serde(rename_all = "kebab-case")]
pub enum ProgressBar {
    #[default]
    Off,
    Fancy,
    Ascii,
}

impl fmt::Display for ProgressBar {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::Off => "off",
                Self::Fancy => "fancy",
                Self::Ascii => "ascii",
            }
        )
    }
}

#[derive(
    Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Debug, Hash, Default, clap::ValueEnum,
)]
#[serde(rename_all = "kebab-case")]
pub enum OutputFormat {
    Concise,
    #[default]
    Full,
    Json,
    JsonLines,
    Junit,
    Grouped,
    Github,
    Gitlab,
    Pylint,
    Rdjson,
    Azure,
    Sarif,
}

impl fmt::Display for OutputFormat {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Concise => write!(f, "concise"),
            Self::Full => write!(f, "full"),
            Self::Json => write!(f, "json"),
            Self::JsonLines => write!(f, "json_lines"),
            Self::Junit => write!(f, "junit"),
            Self::Grouped => write!(f, "grouped"),
            Self::Github => write!(f, "github"),
            Self::Gitlab => write!(f, "gitlab"),
            Self::Pylint => write!(f, "pylint"),
            Self::Rdjson => write!(f, "rdjson"),
            Self::Azure => write!(f, "azure"),
            Self::Sarif => write!(f, "sarif"),
        }
    }
}

#[derive(Debug, Copy, Clone, Hash, is_macro::Is)]
pub enum FixMode {
    Generate,
    Apply,
    #[allow(dead_code)]
    Diff,
}

/// `display_settings!` is a macro that can display and format struct fields in a readable,
/// namespaced format. It's particularly useful at generating `Display` implementations
/// for types used in settings.
///
/// # Example
/// ```
/// use std::fmt;
/// use fortitude::display_settings;
/// #[derive(Default)]
/// struct Settings {
///     option_a: bool,
///     sub_settings: SubSettings,
///     option_b: String,
/// }
///
/// struct SubSettings {
///     name: String
/// }
///
/// impl Default for SubSettings {
///     fn default() -> Self {
///         Self { name: "Default Name".into() }
///     }
///
/// }
///
/// impl fmt::Display for SubSettings {
///     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
///         display_settings! {
///             formatter = f,
///             namespace = "sub_settings",
///             fields = [
///                 self.name | quoted
///             ]
///         }
///         Ok(())
///     }
///
/// }
///
/// impl fmt::Display for Settings {
///     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
///         display_settings! {
///             formatter = f,
///             fields = [
///                 self.option_a,
///                 self.sub_settings | nested,
///                 self.option_b | quoted,
///             ]
///         }
///         Ok(())
///     }
///
/// }
///
/// const EXPECTED_OUTPUT: &str = r#"option_a = false
/// sub_settings.name = "Default Name"
/// option_b = ""
/// "#;
///
/// fn main() {
///     let settings = Settings::default();
///     assert_eq!(format!("{settings}"), EXPECTED_OUTPUT);
/// }
/// ```
#[macro_export]
macro_rules! display_settings {
    (formatter = $fmt:ident, namespace = $namespace:literal, fields = [$($settings:ident.$field:ident $(| $modifier:tt)?),* $(,)?]) => {
        {
            const _PREFIX: &str = concat!($namespace, ".");
            $(
                display_settings!(@field $fmt, _PREFIX, $settings.$field $(| $modifier)?);
            )*
        }
    };
    (formatter = $fmt:ident, fields = [$($settings:ident.$field:ident $(| $modifier:tt)?),* $(,)?]) => {
        {
            const _PREFIX: &str = "";
            $(
                display_settings!(@field $fmt, _PREFIX, $settings.$field $(| $modifier)?);
            )*
        }
    };
    (@field $fmt:ident, $prefix:ident, $settings:ident.$field:ident | debug) => {
        writeln!($fmt, "{}{} = {:?}", $prefix, stringify!($field), $settings.$field)?;
    };
    (@field $fmt:ident, $prefix:ident, $settings:ident.$field:ident | path) => {
        writeln!($fmt, "{}{} = \"{}\"", $prefix, stringify!($field), $settings.$field.display())?;
    };
    (@field $fmt:ident, $prefix:ident, $settings:ident.$field:ident | quoted) => {
        writeln!($fmt, "{}{} = \"{}\"", $prefix, stringify!($field), $settings.$field)?;
    };
    (@field $fmt:ident, $prefix:ident, $settings:ident.$field:ident | globmatcher) => {
        writeln!($fmt, "{}{} = \"{}\"", $prefix, stringify!($field), $settings.$field.glob())?;
    };
    (@field $fmt:ident, $prefix:ident, $settings:ident.$field:ident | nested) => {
        write!($fmt, "{}", $settings.$field)?;
    };
    (@field $fmt:ident, $prefix:ident, $settings:ident.$field:ident | optional) => {
        {
            write!($fmt, "{}{} = ", $prefix, stringify!($field))?;
            match &$settings.$field {
                Some(value) => writeln!($fmt, "{}", value)?,
                None        => writeln!($fmt, "none")?
            };
        }
    };
    (@field $fmt:ident, $prefix:ident, $settings:ident.$field:ident | array) => {
        {
            write!($fmt, "{}{} = ", $prefix, stringify!($field))?;
            if $settings.$field.is_empty() {
                writeln!($fmt, "[]")?;
            } else {
                writeln!($fmt, "[")?;
                for elem in &$settings.$field {
                    writeln!($fmt, "\t{elem},")?;
                }
                writeln!($fmt, "]")?;
            }
        }
    };
    (@field $fmt:ident, $prefix:ident, $settings:ident.$field:ident | map) => {
        {
            use itertools::Itertools;

            write!($fmt, "{}{} = ", $prefix, stringify!($field))?;
            if $settings.$field.is_empty() {
                writeln!($fmt, "{{}}")?;
            } else {
                writeln!($fmt, "{{")?;
                for (key, value) in $settings.$field.iter().sorted_by(|(left, _), (right, _)| left.cmp(right)) {
                    writeln!($fmt, "\t{key} = {value},")?;
                }
                writeln!($fmt, "}}")?;
            }
        }
    };
    (@field $fmt:ident, $prefix:ident, $settings:ident.$field:ident | set) => {
        {
            use itertools::Itertools;

            write!($fmt, "{}{} = ", $prefix, stringify!($field))?;
            if $settings.$field.is_empty() {
                writeln!($fmt, "[]")?;
            } else {
                writeln!($fmt, "[")?;
                for elem in $settings.$field.iter().sorted_by(|left, right| left.cmp(right)) {
                    writeln!($fmt, "\t{elem},")?;
                }
                writeln!($fmt, "]")?;
            }
        }
    };
    (@field $fmt:ident, $prefix:ident, $settings:ident.$field:ident | paths) => {
        {
            write!($fmt, "{}{} = ", $prefix, stringify!($field))?;
            if $settings.$field.is_empty() {
                writeln!($fmt, "[]")?;
            } else {
                writeln!($fmt, "[")?;
                for elem in &$settings.$field {
                    writeln!($fmt, "\t\"{}\",", elem.display())?;
                }
                writeln!($fmt, "]")?;
            }
        }
    };
    (@field $fmt:ident, $prefix:ident, $settings:ident.$field:ident) => {
        writeln!($fmt, "{}{} = {}", $prefix, stringify!($field), $settings.$field)?;
    };
}
