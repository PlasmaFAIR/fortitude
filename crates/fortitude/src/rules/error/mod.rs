pub mod ioerror;
pub mod syntax_error;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::settings::Settings;
    use crate::test::test_path;

    #[test_case(Rule::SyntaxError, Path::new("E001.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("error").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
