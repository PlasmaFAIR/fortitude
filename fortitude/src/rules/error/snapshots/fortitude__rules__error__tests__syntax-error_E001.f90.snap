---
source: fortitude/src/rules/error/mod.rs
expression: diagnostics
snapshot_kind: text
---
./resources/test/fixtures/error/E001.f90:4:23: E001 Syntax error
  |
2 |   implicit none
3 | contains
4 |   subroutine bar(x, y, )
  |                       ^ E001
5 |     integer, intent(in) :: x, y
6 |   end subroutine bar
  |
